### ISSUE TITLE
Insecure configuration for OpenSSH found

### ISSUE THREAD

**Issue by fshmcallister** [WoTTsecurity/wott-io#206]
We found insecure configuration issues with OpenSSH on {devices}. To improve the security posture of your node, please consider making the following changes: {changes}

**Comment by Menmarin** [comment]
## Title
Insecure OpenSSH config detected


## TL:DR 
We have detected security issues in the configuration of OpenSSH on [DEVICES]. Please consider making the recommended changes to the config file to improve the security posture of your node.


## FAQ Version
As you probably know the SSH (Secure Shell) is a method to remote login to a server. SSH is usually secure, however if used with the default settings, can be vulnerable to attacks from unwanted parties.

By default, the SSH config file is set up so that everything is basic and easy to follow. Whilst this is a great way for new developers to get used to setting up SSH servers, it paves the way for vulnerabilities to be exploited. Simple vulnerabilities such as plain text passwords and giving users more access than they need. For more information on how to secure your OpenSSH configuration, please follow [this step-by-step guide](https://wott.io/blog/tutorials/2019/10/25/audit-openssh-security).

**Comment by Menmarin** [comment]
I feel like I can add more the the FAQ but the article I wrote covers all that, so should I just take some from the article and put it in the FAQ or is linking the article enough?

**Comment by vpetersson** [comment]
@fshmcallister / @Menmarin - this one needs a bit rework as we've gone ahead and refactored this in the back-end to make it fit the workflow. We need to break the recommended actions for OpenSSh into three different messages:

 * PermitRootLogin
 * PasswordAuthentication 
 * AllowAgentForwarding

Each need its own version.

**Comment by fshmcallister** [comment]
**Title**
Root login enabled on OpenSSH

**TL;DR**
It appears that you have Root login enabled in your OpenSSH config which may leave your entire machine vulnerable to attack. Please consider disabling this in your config file `/etc/ssh/ssh_config`.

**FAQ**
SSH (Secure Shell) is a method to remote login to a server. SSH is usually secure, however if used with the default settings, can be vulnerable to attacks from unwanted parties. By default, the SSH config file is preset with some values allowing for new developers to get used to setting up SSH servers. However, this often exposes vulnerabilities by not being restrictive enough. 

Here, `PermitRootLogin` is enabled. A common attack is brute forcing the login of superuser and thereby compromising the entire machine rather than on a user-by-user basis. First, ensure you are able to use an SSH login on a non-root user. Once you have confirmed this, disable the Root login by changing the `/etc/ssh/ssh_config` file like so:

```
PermitRootLogin no
```

**Comment by fshmcallister** [comment]
**Title**
Password authentication enabled for OpenSSH

**TL;DR**
It appears that you have Password authentication enabled in your OpenSSH config which may leave your entire machine vulnerable to attack. Please consider removing this in your config file `/etc/ssh/ssh_config`.

**FAQ**
SSH (Secure Shell) is a method to remote login to a server. SSH is usually secure, however if used with the default settings, can be vulnerable to attacks from unwanted parties. By default, the SSH config file is preset with some values allowing for new developers to get used to setting up SSH servers. However, this often exposes vulnerabilities by not being restrictive enough. 

Here, `PasswordAuthentication` is enabled. Passwords can easily be exploited and brute-forced due to human error and is less secure than identifying an accessor via cryptographic identity. Where possible, it is ideal to completely remove the need for passwords as no password is truly safe/uncrackable. Edit the `/etc/ssh/ssh_config` file like so:

```
PasswordAuthentication no
```

For more information on how to secure OpenSSH, see our larger tutorial [here](https://wott.io/blog/tutorials/2019/10/25/audit-openssh-security)

**Comment by fshmcallister** [comment]
**Title**
Agent forwarding enabled for OpenSSH

**TL;DR**
It appears that you have Agent forwarding enabled in your OpenSSH config which may leave your entire machine vulnerable to attack. Please consider disabling this in your config file `/etc/ssh/ssh_config`.

**FAQ**
SSH (Secure Shell) is a method to remote login to a server. SSH is usually secure, however if used with the default settings, can be vulnerable to attacks from unwanted parties. By default, the SSH config file is preset with some values allowing for new developers to get used to setting up SSH servers. However, this often exposes vulnerabilities by not being restrictive enough. 

Here, `AllowAgentForwarding` is enabled. This means that you can SSH into one host and from there to a different host. This might be the case if you cannot directly access the final host. However, doing so leaves a socket open on your machine. You can essentially be impersonated by the first host by anyone with permission to access that node. It is unnecessarily risky and we recommend you disable this. Edit the `/etc/ssh/ssh_config` file like so:

```
AllowAgentForwarding no
```

For more information on how to secure OpenSSH, see our larger tutorial [here](https://wott.io/blog/tutorials/2019/10/25/audit-openssh-security)

**Comment by a-martynovich** [comment]
@fshmcallister @vpetersson This list is missing `PermitEmptyPasswords` and `Protocol`.

**Comment by vpetersson** [comment]
@fshmcallister Could you tackle that too? Thanks.

**Comment by fshmcallister** [comment]
Will need to be done this evening if that’s okay

**Comment by fshmcallister** [comment]
**Title**
Empty passwords enabled for OpenSSH

**TL;DR**
It appears that you have empty passwords enabled in your OpenSSH config. If enabled, an attacker may be able to access your system without a password.

**FAQ**
SSH (Secure Shell) is a method to remote login to a server. SSH is usually secure, however if used with the default settings, can be vulnerable to attacks from unwanted parties. By default, the SSH config file is preset with some values allowing for new developers to get used to setting up SSH servers. However, this often exposes vulnerabilities by not being restrictive enough. 

We've detected that \`PermitEmptyPasswords\` is enabled. This means a user can make changes or access your information without needing a password, as the password field is optional (can be empty). This is particularly dangerous for example if the root user has an empty password. This is not recommended in any instance, particularly over the internet.

To disable this, access and change the \`/etc/ssh/sshd_config\` file like so:

\`\`\`
PermitEmptyPasswords no
\`\`\`

For more information on how to secure OpenSSH, see our larger tutorial \[here\](https://wott.io/blog/tutorials/2019/10/25/audit-openssh-security)

**Code Block**

Not applicable.

**Comment by fshmcallister** [comment]
**Title**
Insecure protocol option enabled for OpenSSH

**TL;DR**
It appears that you have enabled protocol 1 for OpenSSH. This protocol is outdated and insecure, consider exclusively using protocol 2. 

**FAQ**
SSH (Secure Shell) is a method to remote login to a server. SSH is usually secure, however if used with the default settings, can be vulnerable to attacks from unwanted parties. By default, the SSH config file is preset with some values allowing for new developers to get used to setting up SSH servers. However, this often exposes vulnerabilities by not being restrictive enough. 

We've detected that the option to use protocol 1 in \`Protocol\` is enabled. Protocol 1 is insecure compared to protocol 2 as it is vulnerable to insertion attacks and is weak cryptographically. By default OpenSSH will opt to use protocol 2 and in failing that, protocol 1. However, the option to use protocol 1 at all is still unsafe. We suggest that you remove the secondary preference for protocol 1 and disable it as an option at all.

To do so, access and change the \`/etc/ssh/sshd_config\` file like so:

\`\`\`
Protocol 2
\`\`\`

For more information on how to secure OpenSSH, see our larger tutorial \[here\](https://wott.io/blog/tutorials/2019/10/25/audit-openssh-security)

**Code Block**

Not applicable.
